src/tls_wrap.cc

```diff
@@ -747,20 +755,35 @@ int TLSWrap::DoWrite(WriteWrap* w,
   crypto::MarkPopErrorOnReturn mark_pop_error_on_return;
 
   int written = 0;
-  if (count != 1) {
+
+  // It is common for zero length buffers to be written,
+  // don't copy data if there there is one buffer with data
+  // and one or more zero length buffers.
+  // _http_outgoing.js writes a zero length buffer in
+  // in OutgoingMessage.prototype.end.  If there was a large amount
+  // of data supplied to end() there is no sense allocating
+  // and copying it when it could just be used.
+
+  if (nonempty_count != 1) {
     data = env()->AllocateManaged(length);
     size_t offset = 0;
     for (i = 0; i < count; i++) {
       memcpy(data.data() + offset, bufs[i].base, bufs[i].len);
       offset += bufs[i].len;
     }
+
+    crypto::NodeBIO::FromBIO(enc_out_)->set_allocate_hint(length);
     written = SSL_write(ssl_.get(), data.data(), length);
   } else {
     // Only one buffer: try to write directly, only store if it fails
-    written = SSL_write(ssl_.get(), bufs[0].base, bufs[0].len);
+
+    crypto::NodeBIO::FromBIO(enc_out_)->set_allocate_hint(bufs[nonempty_i].len);
+    written =
+    SSL_write(ssl_.get(), bufs[nonempty_i].base, bufs[nonempty_i].len);
```

Comment:
tiny style nit:

```suggestion
        SSL_write(ssl_.get(), bufs[nonempty_i].base, bufs[nonempty_i].len);
```

(4 spaces of indentation for statement continuations)

Reply:
That triggers the long line rule I believe.

Reply:
It’s still shorter than the line before in that case, though? :)

Reply:
Fixed it up in the new commit.